mysticeti-core/src/block_manager.rs

```diff
@@ -62,6 +76,17 @@ impl BlockManager {
                 // If we are missing a reference then we insert into pending and update the waiting index
                 if !self.block_store.block_exists(*included_reference) {
                     processed = false;
+
+                    // we inserted the missing reference for the first time and the block has not been
+                    // fetched already.
+                    if !self
```

Comment:
> Is it possible that !self.blocks_pending.contains_key(included_reference) but block_references_waiting.contains_key(included_reference) is true?

yes that's possible and we expect that - practically that happens when we are missing a reference which we haven't fetched yet at all.

> Why not put this logic in the if !self.blocks_pending.contains_key(included_reference) { block below?

doing that wouldn't allow us then to properly do the check `block_references_waiting.contains_key(included_reference)` - as we already to that on the line 83 - because it's important to return the missing blocks only the first time that we encounter them as missing (so we avoid a fetch amplification later). We could do some small refactoring to acommodate that but I thought it would be cleaner to have this logic completely separate for now.